ansible/playbooks/adhoc/hotfix/roles/update_master/tasks/main.yml

```diff
@@ -0,0 +1,70 @@
+---
+
+# Default to False before attempting upgrade
+- set_fact: 
+    update_successful:    False
+    downgrade_successful: False
+  when: "{{ update_needed == True }}" 
+
+- name: Update atomic-openshift-master
+  yum:
+    name: "{{ rpm_wanted }}"
+    state: present
+  when: "{{ update_needed == True }}" 
+
+- name: Check if update is successful
+  command: "rpm -qa atomic-openshift"
+  register: post_upgrade
+  changed_when: False
+
+# ansible will return a '17' if a match is found. '-1' means 'not found'
+- name: Set update_successful=True if upgrade succeeded
+  set_fact: 
+    update_successful: True
+  when: "{{ post_upgrade.stdout.find(version_wanted) != -1 }}"
+
+- name: Upgrade failed, so try a downgrade
+  command: "yum -y downgrade atomic-openshift-{{ version_wanted }} atomic-openshift-clients-{{ version_wanted }} atomic-openshift-master-{{ version_wanted }} atomic-openshift-node-{{ version_wanted }} tuned-profiles-atomic-openshift-node-{{ version_wanted }} atomic-openshift-sdn-ovs-{{ version_wanted }}"
+  when: "{{ update_successful == False }}"
```

Comment:
I'd prefer it if we had a separate "rollback.yml" playbook rather than automatically attempting to fix a failed upgrade cluster / host.

I prefer this because I want the rollback gated on a human making the decision of what's best. For instance, we might decide to just push forward, rather than roll back.

It will also allow us to run the rollback if, for instance, the upgrade succeeded _but_ there's some catastrophic bug that forces us to roll back to the previous version.

Also, here, you're specifying the version of the packages to downgrade to. And, it's the same version of the packages that we just tried to upgrade to.

That doesn't seem right. It should be the previous version, right?

Lastly, I'd prefer it if we used "yum history undo" as our rollback mechanism because:
1) it'll ensure we're back on the same rpm versions we were on before (we won't accidentally upgrade the cluster to a higher 3.2.1.x version than what it was before).
2) it'll ensure that any other RPMs that got updated during the attempted upgrade are also set back to their original versions (like maybe docker or anything else that are deps from these packages)

We'll need to ensure that we only run the "yum history undo" on the upgrade transaction (we wouldn't want to undo a security update, for instance).